src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs

```diff
@@ -327,19 +425,349 @@ public async Task StartTestRunWithCustomHostAsync(IEnumerable<TestCase> testCase
                     TestCases = testCases.ToList(),
                     RunSettings = runSettings,
                     DebuggingEnabled = customHostLauncher.IsDebug,
-                    TestPlatformOptions = options
+                    TestPlatformOptions = options,
+                    TestSessionInfo = testSessionInfo
                 },
                 runEventsHandler,
                 customHostLauncher);
         }
 
+        /// <inheritdoc/>
+        public TestSessionInfo StartTestSession(
+            IList<string> sources,
+            string runSettings,
+            TestPlatformOptions options,
+            ITestSessionEventsHandler eventsHandler,
+            ITestHostLauncher testHostLauncher)
+        {
+            if (EqtTrace.IsInfoEnabled)
+            {
+                EqtTrace.Info("VsTestConsoleRequestSender.StartTestSession: Starting test session.");
+            }
+
+            try
+            {
+                var payload = new StartTestSessionPayload
+                {
+                    // TODO (copoiena): When sharing the test host between test discovery and test
+                    // execution, should we use the test host launcher to launch it ? What side
+                    // effects does this have ?
+                    //
+                    // This is useful for profiling and maybe for launching hosts other than the
+                    // ones managed by us (i.e., the default host and the dotnet host), examples
+                    // including UWP and other hosts that don't implement the ITestRuntimeProvider2
+                    // interface and/or are not aware of the possibility of attaching to an already
+                    // running process.
+                    Sources = sources,
+                    RunSettings = runSettings,
+                    HasCustomHostLauncher = testHostLauncher != null,
+                    IsDebuggingEnabled = (testHostLauncher != null)
+                        ? testHostLauncher.IsDebug
+                        : false,
+                    TestPlatformOptions = options
+                };
+
+                this.communicationManager.SendMessage(
+                    MessageType.StartTestSession,
+                    payload,
+                    this.protocolVersion);
+
+                while (true)
+                {
+                    var message = this.TryReceiveMessage();
+
+                    switch (message.MessageType)
```

Comment:
Yes, unfortunately they're mostly the same, the exception being the async parts. We could attempt splitting the repeating code in multiple methods but the way I'm thinking of if will end up hurting readability because the method encodes one large state machine.

Do you have any thoughts on how we can do this in a more elegant way ?